### PR TITLE
fix: Default value is ignored when loading params from configmap. Fixes #8262

### LINKED ISSUE #8262: ConfigmapRef does not respect optional and default keyword
## Checklist

 

* [x] Double-checked my configuration.
* [ ] Tested using the latest version.
* [x] Used the Emissary executor.

## Summary

What happened/what you expected to happen?
`Optional` and `default` keywords are not respected when using  `configMapKeyRef`.
Workflow errors while trying to reference a key that does not exist.

What version are you running?
v3.2.9
Emissary Executor

## Diagnostics

Paste the smallest workflow that reproduces the bug. We must be able to run the workflow.

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: retry-sample
  labels:
    workflows.argoproj.io/configmap-type: Parameter
data:
  dummy: "{}"
  dummy1: ""
```

```yaml
apiVersion: argoproj.io/v1alpha1
kind: WorkflowTemplate
metadata:
  name: retry-sample
spec:
  entrypoint: main
  templates:
    - name: main
      inputs:
        parameters:
          - name: message
            valueFrom:
              default: "hello"
              configMapKeyRef:
                name: retry-sample
                key: dummy-missing
                optional: true
      container:
        image: docker/whalesay
        command: [cowsay]
        args: ["{{inputs.parameters.message}}"]
```

```bash
# Logs from the workflow controller:
kubectl logs -n argo deploy/workflow-controller | grep ${workflow} 

# If the workflow's pods have not been created, you can skip the rest of the diagnostics.

# The workflow's pods that are problematic:
kubectl get pod -o yaml -l workflows.argoproj.io/workflow=${workflow},workflow.argoproj.io/phase!=Succeeded

# Logs from in your workflow's wait container, something like:
kubectl logs -c wait -l workflows.argoproj.io/workflow=${workflow},workflow.argoproj.io/phase!=Succeeded
```

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by terrytangyuan**
Fixes https://github.com/argoproj/argo-workflows/issues/8262.

Signed-off-by: Yuan Tang <terrytangyuan@gmail.com>

### CODE REVIEW

File: workflow/common/util.go [hunk 1]
```diff
@@ -151,9 +151,14 @@ func ProcessArgs(tmpl *wfv1.Template, args wfv1.ArgumentsProvider, globalParams,
 
 				cmValue, err := GetConfigMapValue(configMapInformer, namespace, cmName, cmKey)
 				if err != nil {
-					return nil, errors.Errorf(errors.CodeBadRequest, "unable to retrieve inputs.parameters.%s from ConfigMap: %s", inParam.Name, err)
+					if inParam.ValueFrom.Default != nil {
```
Comment (alexec) [inline workflow/common/util.go:154]: should we make sure `err` is IsNotFound?
what about test?

Comment (terrytangyuan) [reply workflow/common/util.go:154]: > Default specifies a value to be used if retrieving the value from the specified source fails
Based on the original purpose from the doc-string, I think we want to have default no matter what specific error we are getting. 

I'll add tests.

Comment (alexec) [reply workflow/common/util.go:154]: I'm not sure about this, should a transient error result in my workflow running with unexpected parameters?

Comment (terrytangyuan) [reply workflow/common/util.go]: That makes sense. Added a check for IsNotFound as well as a test.

**Review by alexec — APPROVED** [review]

**Pull request merged**